### PR TITLE
Introduce warnings to HBO retry mechanism

### PR DESCRIPTION
**Pull request by abhinavmuk04**
## Description
Introduce warnings to HBO retry mechanism

## Motivation and Context
When running a query and seeing it, either through the CLI ro through the Presto Query Lookup tool, warning messages will help clarify for the user what is happening

## Impact
Since there are many moving parts with the users and the query retry mechanism, alerting the user at each big step will be helpful. After seeing their query fail and then retry the warning message will provide clarity and more information regarding the status of the retry as well.

## Test Plan
E2E tests in a test cluster

## Contributor checklist

- [x] Please make sure your submission complies with our [development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), [commit message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests), and [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).
- [x] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [x] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [x] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [x] Adequate tests were added if applicable.
- [x] CI passed.

## Release Notes
Please follow [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) and fill in the release notes below.

If release note is NOT required, use:

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/server/protocol/Query.java [hunk 1]
```diff
@@ -399,19 +401,31 @@ private synchronized QueryResults getNextResultWithRetry(long token, UriInfo uri
         boolean hasNotRetried = queryManager.getQueryRetryCount(queryId) < 1;
 
         if (historyBasedOptimizationEnabled && hasNotRetried && retryConditionsMet(queryResults) && retryQueryWithHistoryBasedOptimizationEnabled(session)) {
+            WarningCode warningFromOriginalQuery = new WarningCode(700, "Your query is being retried");
```
Comment (kaikalur) [inline presto-main/src/main/java/com/facebook/presto/server/protocol/Query.java:404]: Don't we use constants for this WarningCode or something?

File: presto-main/src/main/java/com/facebook/presto/server/protocol/Query.java [hunk 2]
```diff
@@ -399,19 +401,31 @@ private synchronized QueryResults getNextResultWithRetry(long token, UriInfo uri
...
             previousQueryFailureError = Optional.of(queryResults.getError());
         }
         else if (queryManager.getQueryRetryCount(queryId) == 1 && retryQueryWithHistoryBasedOptimizationEnabled(session) && retryConditionsMet(queryResults) && historyBasedOptimizationEnabled) {
             Optional<Integer> currentTopLevelPlanHash = getCurrentTopLevelPlanHash();
 
+            WarningCode warningFromRetryQuery = new WarningCode(701, "This is your retried query");
+
             if (previousQueryTopLevelPlanHash.isPresent() && currentTopLevelPlanHash.isPresent() && currentTopLevelPlanHash.equals(previousQueryTopLevelPlanHash)
                     || (!previousQueryTopLevelPlanHash.isPresent() && !currentTopLevelPlanHash.isPresent())) {
+                String retryQuerySamePlanMessage = "Since the plan hashes did not change, your retry query will not execute";
```
Comment (kaikalur) [inline presto-main/src/main/java/com/facebook/presto/server/protocol/Query.java:420]: Planhash is too low level to surface to users. Reword it to say something like - HBO did not find better plan aborting retry

File: presto-main/src/main/java/com/facebook/presto/server/protocol/Query.java [hunk 3]
```diff
@@ -399,19 +401,31 @@ private synchronized QueryResults getNextResultWithRetry(long token, UriInfo uri
...
+                session.getWarningCollector().add(new PrestoWarning(warningFromRetryQuery, retryQuerySamePlanMessage));
+
                 queryManager.failQuery(queryId, new PrestoException(GENERIC_INTERNAL_ERROR, "Since the plan hashes did not change, your retry query will not execute." +
                         "Your original error was " + previousQueryFailureError.get() + ". Original QueryId: " + originalBeforeRetryQueryId +
                         ". Retry QueryId: " + queryId));
             }
+            String retryQueryMessage = "This query is a retry of the failed query: " + originalBeforeRetryQueryId + ".";
```
Comment (kaikalur) [inline presto-main/src/main/java/com/facebook/presto/server/protocol/Query.java:427]: I don't think we put period at the end of our warnings - double check

**Review by kaikalur — CHANGES_REQUESTED** [review]

File: presto-main/src/main/java/com/facebook/presto/server/protocol/Query.java (hunk 1 above)
Comment (abhinavmuk04) [reply presto-main/src/main/java/com/facebook/presto/server/protocol/Query.java:404]: Yeah you are right, I'll make it an enum

File: presto-main/src/main/java/com/facebook/presto/server/protocol/Query.java (hunk 3 above)
Comment (abhinavmuk04) [reply presto-main/src/main/java/com/facebook/presto/server/protocol/Query.java:427]: https://www.internalfb.com/intern/presto/query/?query_id=20240724_225325_00679_vwf49#dr_presto in this warning we do, not sure if you count this

Comment (abhinavmuk04) [reply presto-main/src/main/java/com/facebook/presto/server/protocol/Query.java:424]: Think in the message we have them just not in the title

File: presto-main/src/main/java/com/facebook/presto/server/protocol/Query.java [hunk 4]
```diff
@@ -408,10 +414,15 @@ else if (queryManager.getQueryRetryCount(queryId) == 1 && retryQueryWithHistoryB
 
             if (previousQueryTopLevelPlanHash.isPresent() && currentTopLevelPlanHash.isPresent() && currentTopLevelPlanHash.equals(previousQueryTopLevelPlanHash)
                     || (!previousQueryTopLevelPlanHash.isPresent() && !currentTopLevelPlanHash.isPresent())) {
-                queryManager.failQuery(queryId, new PrestoException(GENERIC_INTERNAL_ERROR, "Since the plan hashes did not change, your retry query will not execute." +
+                String retryQuerySamePlanMessage = "HBO did not find better plan aborting retry";
+                session.getWarningCollector().add(new PrestoWarning(CURRENT_QUERY_IS_RETRY_QUERY, retryQuerySamePlanMessage));
+
+                queryManager.failQuery(queryId, new PrestoException(GENERIC_INTERNAL_ERROR, "HBO did not find better plan aborting retry." +
```
Comment (kaikalur) [inline presto-main/src/main/java/com/facebook/presto/server/protocol/Query.java:420]: Don't make this internal error - maybe just make it a info/logging if the user doesn't see it - which they shouldn't

File: presto-spi/src/main/java/com/facebook/presto/spi/StandardWarningCode.java [hunk 5]
```diff
@@ -25,7 +25,9 @@ public enum StandardWarningCode
     MULTIPLE_ORDER_BY(0x0000_0007),
     DEFAULT_SAMPLE_FUNCTION(0x0000_0008),
     SAMPLED_FIELDS(0x0000_0009),
-    MULTIPLE_TABLE_METADATA(0x0000_0010)
+    MULTIPLE_TABLE_METADATA(0x0000_0010),
+    QUERY_RETRY_STARTED(0x0000_0011),
+    CURRENT_QUERY_IS_RETRY_QUERY(0x0000_0011)
```
Comment (kaikalur) [inline presto-spi/src/main/java/com/facebook/presto/spi/StandardWarningCode.java:30]: use different code

**Review by kaikalur — CHANGES_REQUESTED** [review]

File: presto-main/src/main/java/com/facebook/presto/server/protocol/Query.java (hunk 4 above)
Comment (abhinavmuk04) [reply presto-main/src/main/java/com/facebook/presto/server/protocol/Query.java:420]: @kaikalur https://github.com/prestodb/presto/pull/23372

Comment (kaikalur) [reply presto-main/src/main/java/com/facebook/presto/server/protocol/Query.java:420]: Not sure what that issue says but in general we shouldnt use internal error as it's counted as something went wrong. Here nothing went wrong. So I suggest just info logging it. User doesn't need to know

**Review by kaikalur — CHANGES_REQUESTED** [review]
